### PR TITLE
Detect mips CPUs in ./configure

### PR DESCRIPTION
**Pull request by infinity0**
This mirrors existing logic already in src/bootstrap/bootstrap.py

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -388,6 +388,7 @@ msg "inspecting environment"
 
 CFG_OSTYPE=$(uname -s)
 CFG_CPUTYPE=$(uname -m)
+ENDIAN=$(printf '\1' | od -dAn)
```
Comment (xen0n) [inline configure]: Here `od` isn't `need_cmd`'d, thus unsafe to rely on. You could refer to the [`rustup-init.sh`] implementation for an alternative that only requires `head` and `tail`.

[`rustup-init.sh`]: https://github.com/rust-lang-nursery/rustup.rs/blob/1bedbb1b81a9a4cf6c8927db27a957e8b1fa18bb/rustup-init.sh#L125-L142

**Review by xen0n — COMMENTED** [review]
Thanks for spotting this! I forgot to touch `configure` in my MIPS host build PR because the MIPS64 platform I was working on is rustbuild-only. Looks good overall!

**Comment by infinity0** [comment]
Welcome! In the meantime I spotted some more things which are missing:

- mk/cfg/mips{,el}-unknown-linux-gnu.mk: doesn't define CFG_RUN_TARG_mips{,el}-unknown-linux-gnu
- mk/cfg/mips64{,el}-unknown-linux-gnuabi64.mk are empty

I'll fix these as well and add them to this PR.

In the long run, does the rust team have any plans to unify the Makefiles and rustup/rustbuild? Seems like unnecessary work to have to maintain both of these...

File: configure (hunk 1 above)
Comment (infinity0) [inline configure]: `od` is [part of](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/od.html) POSIX (including the `-d` and `-A n` options), just as `head` and `tail` are. Is that good enough? If not, I can do your suggestion.

**Comment by xen0n** [comment]
@infinity0 Please refrain from touching the makefiles, they're [slated to be deleted pretty soon][deprecated]!

[deprecated]: https://internals.rust-lang.org/t/proposal-for-promoting-rustbuild-to-official-status/4368

File: configure (hunk 1 above)
Comment (xen0n) [inline configure]: I don't know, but some exotic environments indeed doesn't have `od`; a quick Googling reveals OpenWRT and Maemo and the such. I don't know how many people are compiling Rust *on* these platforms though...

**Comment by infinity0** [comment]
Ouch, OK. Is there a way to not use Cargo to build rust? That is going to make bootstrapping new architectures much much more difficult.

**Comment by xen0n** [comment]
@infinity0 No, bootstraping will be arguably *easier*, since you no longer have to deal with multiple different build tools with rustbuild. Look at [the way `rustc` is `configure`'d for the non-x86 hosts][cross-rustbuild], pretty much *all* of the work you're expected to do for bootstrapping is now concentrated in the codebase itself. See #38314 for a feel of it; bootstrapping is lighter than ever IMO.

[cross-rustbuild]: https://buildbot.rust-lang.org/builders/nightly-dist-rustc-cross-rustbuild-linux/builds/84/steps/configure/logs/stdio

**Comment by infinity0** [comment]
Er, I am doing bootstrapping *literally right now*, and I can tell you it will be harder. All of the previous tools like `make` etc are already present for other architectures; cargo is not. Large cyclic dependencies also increase the amount of binary non-source code that needs to be trusted when bootstrapping a given whole distribution. From this perspective, it is better for rust to depend on 100 other build tools (that eventually depend on a small set of cyclic-dependency base build tools) than to be cyclicly dependent with cargo. I can appreciate the code is nicer to review than the previous Makefiles, but it would be nice if this could be done in a way that does not hard-depend on cargo. I made a concrete suggestion in #38652 and we can continue the discussion there.

**Comment by sylvestre** [comment]
:+1: to what @infinity0 just said.

File: configure (hunk 1 above)
Comment (alexcrichton) [inline configure]: In order to minimize the possible impact of this change, could this move into just the mips block below?

**Comment by alexcrichton** [comment]
This PR looks good to me, thanks @infinity0!

**Comment by infinity0** [comment]
OK, I've moved `ENDIAN` down into the mips-specific block. I also added another commit that was necessary to compile the tests. Due to other problems (#39013, #39014) I couldn't test these all the way through to completion, however I notice that these files:
~~~~
src/tools/compiletest/src/util.rs
src/test/run-make/atomic-lock-free/Makefile
~~~~
mention `mips` but not `mips64`. Is that intentional or do they also need to be fixed?

**Comment by alexcrichton** [comment]
@bors: r+ 

Thanks! It's ok to not update those unless necessary.

**Pull request merged**